src/session.rs

```diff
@@ -0,0 +1,373 @@
+                return Some(format!("Basic {}", base64::encode(raw_auth)));
+            } else if auth_type.as_str() == "bearer" {
+                return Some(format!("Bearer {}", raw_auth));
+            }
+        }
+
+        None
+    }
+
+    pub fn save_bearer_auth(&mut self, token: String) {
+        self.content.auth = Auth {
+            auth_type: Some("bearer".into()),
+            raw_auth: Some(token),
+        }
+    }
+
+    pub fn save_basic_auth(&mut self, username: String, password: Option<String>) {
+        self.content.auth = Auth {
+            auth_type: Some("basic".into()),
+            raw_auth: Some(format!("{}:{}", username, password.unwrap_or("".into()))),
+        }
+    }
+
+    pub fn cookies(&self) -> Vec<cookie_crate::Cookie> {
+        let mut cookies = vec![];
+        for (name, c) in self.content.cookies.iter() {
+            let mut cookie_builder = cookie_crate::Cookie::build(name, &c.value);
+            if let Some(expires) = c.expires {
+                cookie_builder =
+                    cookie_builder.expires(time::OffsetDateTime::from_unix_timestamp(expires));
+            }
+            if let Some(ref path) = c.path {
+                cookie_builder = cookie_builder.path(path);
+            }
+            if let Some(secure) = c.secure {
+                cookie_builder = cookie_builder.secure(secure);
+            }
+            cookies.push(cookie_builder.finish());
+        }
+        cookies
+    }
+
+    pub fn save_cookies(&mut self, cookies: Vec<cookie_crate::Cookie>) {
+        self.content.cookies.clear();
+        for cookie in cookies {
+            self.content.cookies.insert(
+                cookie.name().into(),
+                Cookie {
+                    value: cookie.value().into(),
+                    expires: cookie
+                        .expires()
+                        .and_then(|v| v.datetime())
+                        .map(|v| v.unix_timestamp()),
+                    path: cookie.path().map(Into::into),
+                    secure: cookie.secure(),
+                },
+            );
+        }
+    }
+
+    pub fn persist(&self) -> Result<()> {
+        if !self.path.exists() || !self.read_only {
+            if let Some(parent_path) = self.path.parent() {
+                fs::create_dir_all(parent_path)?;
+            }
+            let mut session_file = fs::File::create(&self.path)?;
+            let formatter = serde_json::ser::PrettyFormatter::with_indent(b"    ");
+            let mut ser = serde_json::Serializer::with_formatter(&mut session_file, formatter);
+            self.content.serialize(&mut ser)?;
+            session_file.write_all(b"\n")?;
+        }
+        Ok(())
+    }
+}
+
+fn is_path(value: &OsString) -> bool {
+    value.to_string_lossy().contains(std::path::is_separator)
+}
+
+fn config_dir() -> Option<PathBuf> {
+    if test_mode() {
+        std::env::var_os("XH_TEST_CONFIG_DIR").map(Into::into)
```

Comment:
It might be good to let users customize their config directory. Should we rename this to `XH_CONFIG_DIR` so it similar to HTTPie's `HTTPIE_CONFIG_DIR`? or maybe we should have both i.e `XH_CONFIG_DIR` and `XH_TEST_CONFIG_DIR`?

Reply:
I wasn't aware of `$HTTPIE_CONFIG_DIR`, having our own version of that makes sense.

I can't offhand think of a reason to also separately have `$XH_TEST_CONFIG_DIR`.